### PR TITLE
Unverified publishers in panel can now have status refreshed

### PR DESCRIPTION
**Pull request by jasonrsadler**
Resolves brave/brave-browser#3339
Brave-UI PR: https://github.com/brave/brave-ui/pull/438
## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests && npm run test-security`) on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Test Plan:

1. Start Brave, clean profile and enable rewards.
2. Go to an unverified publisher and open panel.
3. Ensure that 'Check Again...' link appears and is clickable
4. Ensure that when 'Check Again...' link is clicked, a spinner appears briefly, then spinner and link disappear
5. Close panel and reopen, and ensure that link is available again

--------------

1. Start Brave, clean profile and enable rewards, then close browser.
2. Go to profile/default directory and open `publishers_list` file.
3. Remove a verified publisher and save and close.
4. Restart Brave and go to the web site of the publisher that was removed from the file.
5. Open the panel and ensure publisher is shown as unverified.
6. Click the 'Check Again...' link and ensure the publisher is now verified
7. Repeat the above with a youtube and twitch publisher

---------------

1. Start Brave, clean profile and enable rewards, then close browser.
2. Go to profile/default directory and open `publishers_list` file.
3. Remove a verified publisher and save and close.
4. Restart Brave and go to the web site of the publisher that was removed from the file.
5. Open the panel and ensure publisher is shown as unverified.
6. Disconnect internet
7. Click the 'Check Again...' link and ensure the publisher status does not change
8. Reconnect internet and click the 'Check Again...' link and ensure publisher is now verified

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

### CODE REVIEW

File: components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx [hunk 1]
```diff
@@ -335,8 +336,19 @@ export class Panel extends React.Component<Props, State> {
     }
   }
 
+  refreshPublisher = () => {
+    const publisher: RewardsExtension.Publisher | undefined = this.getPublisher()
+    const publisherKey = publisher && publisher.publisher_key
+    chrome.tabs.query({
```
Comment (NejcZdovc) [inline components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx]: you don't need to query tab, you can get window id in `this.state.windowId`

File: vendor/bat-native-ledger/src/bat/ledger/internal/bat_publishers.cc [hunk 2]
```diff
@@ -864,4 +864,32 @@ void BatPublishers::onPublisherBanner(
   callback(std::move(new_banner));
 }
 
+void BatPublishers::RefreshPublisherVerifiedStatus(
+    const std::string& publisher_key,
+    ledger::OnRefreshPublisherVerifiedStatusCallback callback) {
+  bool verified = isVerified(publisher_key);
+  ledger_->GetPublisherInfo(publisher_key,
+      std::bind(&BatPublishers::SetPublisherVerified, this,
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/internal/bat_publishers.cc]: nit: please align params one after another

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc [hunk 3]
```diff
@@ -1364,6 +1372,45 @@ void LedgerImpl::GetConfirmationsHistory(
...
+      std::bind(&LedgerImpl::OnRefreshPublisherVerifiedStatus,
+        this, _1, _2, window_id));
+}
+
+void LedgerImpl::OnRefreshPublisherVerifiedStatus(
+    ledger::Result result,
+    std::unique_ptr<ledger::PublisherInfo> publisher_info,
+    uint64_t window_id) {
+  OnPanelPublisherInfo(
+      ledger::Result::LEDGER_OK,
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc]: result should be actual result and not default OK

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc [hunk 4]
```diff
@@ -1364,6 +1372,45 @@ void LedgerImpl::GetConfirmationsHistory(
...
+}
+
+void LedgerImpl::OnPanelPublisherListDownloaded(
+    int response_status_code,
+    const std::string& response,
+    const std::map<std::string, std::string>& headers,
+    uint64_t window_id,
+    const std::string& publisher_key) {
+  if (response_status_code != 200) {
+
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc]: we need to report this back to the panel that we were not successful

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc [hunk 5]
```diff
@@ -1364,6 +1372,45 @@ void LedgerImpl::GetConfirmationsHistory(
...
+  RefreshPublisherVerifiedStatus(publisher_key,
+      std::bind(&LedgerImpl::OnRefreshPublisherVerifiedStatus,
+        this, _1, _2, window_id));
+}
+
+void LedgerImpl::OnRefreshPublisherVerifiedStatus(
+    ledger::Result result,
+    std::unique_ptr<ledger::PublisherInfo> publisher_info,
+    uint64_t window_id) {
+  OnPanelPublisherInfo(
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc]: We should be using callback mechanism to return info to the panel. This way if something is wrong we can tell panel, something like: `Couldn't download list right now, please try again later.`

File: components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx (hunk 1 above)
Comment (jasonrsadler) [reply components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx]: It's in `this.props.windowId`, but thanks. Good to know.

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc (hunk 4 above)
Comment (jasonrsadler) [reply vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc]: This was meant to be removed. We pass `response_status_code` into `LoadPublishersListCallback` to retry the download. Is this sufficient?

File: components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx (hunk 1 above)
Comment (NejcZdovc) [reply components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx]: sorry, it's in `state` in `app.tsx` and in `props` in `panel.tsx`

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc (hunk 5 above)
Comment (jasonrsadler) [reply vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc]: This has been reworked to return in callback. Let me know if this is what you are looking for.

File: vendor/bat-native-ledger/src/bat/ledger/internal/bat_publishers.cc (hunk 2 above)
Comment (jasonrsadler) [reply vendor/bat-native-ledger/src/bat/ledger/internal/bat_publishers.cc]: Done.

File: components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx (hunk 1 above)
Comment (jasonrsadler) [reply components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx]: Done.

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc (hunk 3 above)
Comment (jasonrsadler) [reply vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc]: Fixed.

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc (hunk 5 above)
Comment (jasonrsadler) [reply vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc]: Returned api callback as discussed in DM

**Comment by jasonrsadler** [comment]
@NejcZdovc Refactored to use api callback mechanism. Important changes to note: The call to native is now done in Panel and the reducer/action code has been removed and the storage.ts vars were no longer needed.

File: browser/extensions/api/brave_rewards_api.cc [hunk 6]
```diff
@@ -291,5 +291,32 @@ ExtensionFunction::ResponseAction BraveRewardsSaveSettingFunction::Run() {
...
+
+ExtensionFunction::ResponseAction BraveRewardsRefreshPublisherFunction::Run() {
+  std::unique_ptr<brave_rewards::RefreshPublisher::Params> params(
+      brave_rewards::RefreshPublisher::Params::Create(*args_));
+
+  Profile* profile = Profile::FromBrowserContext(browser_context());
+  RewardsService* rewards_service =
+    RewardsServiceFactory::GetForProfile(profile);
+  if (!rewards_service) {
+    return RespondNow(OneArgument(std::make_unique<base::Value>(false)));
```
Comment (NejcZdovc) [inline browser/extensions/api/brave_rewards_api.cc]: we should respond with two arguments, like we do in `OnRefreshPublisher`

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 7]
```diff
@@ -2870,4 +2870,22 @@ void RewardsServiceImpl::OnDeleteActivityInfo(
   }
 }
 
+void RewardsServiceImpl::RefreshPublisher(
+    RefreshPublisherCallback callback,
+    const std::string& publisher_key) {
+  if (!Connected())
+    return;
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/rewards_service_impl.cc]: we need to call back here with false and empty string

File: components/brave_rewards/resources/extension/brave_rewards/background/reducers/rewards_panel_reducer.ts [hunk 8]
```diff
@@ -305,6 +305,21 @@ export const rewardsPanelReducer = (state: RewardsExtension.State | undefined, a
       }
       break
     }
+    case types.REFRESH_PUBLISHER: {
+      let publishers: Record<string, RewardsExtension.Publisher> = state.publishers
```
Comment (NejcZdovc) [inline components/brave_rewards/resources/extension/brave_rewards/background/reducers/rewards_panel_reducer.ts]: please add check for empty publisher key, before doing anything else

File: vendor/bat-native-ledger/include/bat/ledger/ledger.h [hunk 9]
```diff
@@ -282,6 +284,9 @@ class LEDGER_EXPORT Ledger {
       const uint64_t to_timestamp_seconds,
       ledger::ConfirmationsHistoryCallback callback) = 0;
   virtual void GetRewardsInternalsInfo(ledger::RewardsInternalsInfo* info) = 0;
+  virtual void RefreshPublisher(
+      ledger::OnRefreshPublisherCallback callback,
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/include/bat/ledger/ledger.h]: nit: can we do first publisher_key and then callback

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.h [hunk 10]
```diff
@@ -438,6 +438,10 @@ class LedgerImpl : public ledger::Ledger,
   scoped_refptr<base::SequencedTaskRunner> GetTaskRunner();
   void GetRewardsInternalsInfo(ledger::RewardsInternalsInfo* info) override;
 
+  void RefreshPublisher(
+      ledger::OnRefreshPublisherCallback callback,
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.h]: nit: can we do first publisher_key and then callback

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.h [hunk 11]
```diff
@@ -501,6 +505,16 @@ class LedgerImpl : public ledger::Ledger,
...
       std::unique_ptr<ledger::PublisherInfo> info);
 
+  void DownloadPublisherList(
+      ledger::LoadURLCallback callback);
+
+  void OnRefreshPublisher(
+      int response_status_code,
+      const std::string& response,
+      const std::map<std::string, std::string>& headers,
+      ledger::OnRefreshPublisherCallback callback,
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.h]: nit: can we do first publisher_key and then callback

File: browser/extensions/api/brave_rewards_api.cc (hunk 6 above)
Comment (jasonrsadler) [reply browser/extensions/api/brave_rewards_api.cc]: Good catch. Fixed

File: components/brave_rewards/browser/rewards_service_impl.cc (hunk 7 above)
Comment (jasonrsadler) [reply components/brave_rewards/browser/rewards_service_impl.cc]: Fixed

File: components/brave_rewards/resources/extension/brave_rewards/background/reducers/rewards_panel_reducer.ts (hunk 8 above)
Comment (jasonrsadler) [reply components/brave_rewards/resources/extension/brave_rewards/background/reducers/rewards_panel_reducer.ts]: Done.

File: vendor/bat-native-ledger/include/bat/ledger/ledger.h (hunk 9 above)
Comment (jasonrsadler) [reply vendor/bat-native-ledger/include/bat/ledger/ledger.h]: Done

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.h (hunk 10 above)
Comment (jasonrsadler) [reply vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.h]: Done.

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.h (hunk 11 above)
Comment (jasonrsadler) [reply vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.h]: Done.

**Comment by NejcZdovc** [comment]
@jasonrsadler could you please rebase to the latest master, not sure why our CI is failing

**Comment by jasonrsadler** [comment]
It's an npm audit 503

**Comment by jasonrsadler** [comment]
rebased

**Comment by NejcZdovc** [comment]
@jasonrsadler yeah Travis is ok, I am wondering why our CI is not working

File: common/extensions/api/brave_rewards.json [hunk 12]
```diff
@@ -563,6 +563,31 @@
...
           }
         ]
+      },
+      {
+        "name": "refreshPublisher",
+        "type": "function",
+        "description": "Refreshes verified status for individual publisher",
+        "parameters": [
+          {
+            "name": "publisherKey",
```
[the remark this answers is not in the document]
Comment (jasonrsadler) [reply common/extensions/api/brave_rewards.json]: One is for the callback

File: common/extensions/api/brave_rewards.json [hunk 13]
```diff
@@ -563,6 +563,31 @@
...
+            "type": "string"
+          },
+          {
+            "type": "function",
+            "name": "callback",
+            "parameters": [
+              {
+			          "name": "verified",
```
[the remark this answers is not in the document]
Comment (jasonrsadler) [reply common/extensions/api/brave_rewards.json]: Fixed

File: package-lock.json [hunk 14]
```diff
@@ -1259,7 +1259,7 @@
         },
         "chalk": {
           "version": "1.1.3",
-          "resolved": "https://registry.npmjs.org/chalk/-/chalk-1.1.3.tgz",
+          "resolved": "http://registry.npmjs.org/chalk/-/chalk-1.1.3.tgz",
```
[the remark this answers is not in the document]
Comment (jasonrsadler) [reply package-lock.json]: Merge artifact. Removed.

File: components/services/bat_ledger/bat_ledger_impl.cc [hunk 15]
```diff
@@ -468,6 +468,15 @@ void BatLedgerImpl::OnGetOneTimeTips(
   delete holder;
 }
 
+void BatLedgerImpl::OnRefreshPublisher(
```
[the remark this answers is not in the document]
Comment (jasonrsadler) [reply components/services/bat_ledger/bat_ledger_impl.cc]: Done

File: common/extensions/api/brave_rewards.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (NejcZdovc) [reply common/extensions/api/brave_rewards.json]: yup my bad

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 16]
```diff
@@ -2922,4 +2922,24 @@ void RewardsServiceImpl::OnDeleteActivityInfo(
   }
 }
 
+void RewardsServiceImpl::RefreshPublisher(
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/rewards_service_impl.cc]: nit: in all other places you use publisher_key first. Can you change it here as well?

Comment (jasonrsadler) [reply components/brave_rewards/browser/rewards_service_impl.cc]: Done.

File: package.json [hunk 17]
```diff
@@ -277,7 +277,7 @@
     "@types/react-redux": "6.0.4",
     "@types/redux-logger": "^3.0.7",
     "awesome-typescript-loader": "^5.2.1",
-    "brave-ui": "github:brave/brave-ui#1b1ee01094d6caa3336dcd42298c4178e6930185",
+    "brave-ui": "github:brave/brave-ui#f2762fd62577aa5e177b462ead9c3c32ef8d915a",
```
Comment (NejcZdovc) [inline package.json]: bump is not needed

Comment (jasonrsadler) [reply package.json]: Yep. Thanks

File: components/brave_rewards/resources/extension/brave_rewards/constants/rewards_panel_types.ts [hunk 18]
```diff
@@ -34,7 +34,8 @@ export const enum types {
   SAVE_RECURRING_TIP = '@@rewards_panel/SAVE_RECURRING_TIP',
   REMOVE_RECURRING_TIP = '@@rewards_panel/REMOVE_RECURRING_TIP',
   ON_RECURRING_TIPS = '@@rewards_panel/ON_RECURRING_TIPS',
-  ON_PUBLISHER_BANNER = '@@rewards_panel/ON_PUBLISHER_BANNER'
+  ON_PUBLISHER_BANNER = '@@rewards_panel/ON_PUBLISHER_BANNER',
+  REFRESH_PUBLISHER = '@@rewards_panel/REFRESH_PUBLISHER'
```
Comment (NejcZdovc) [inline components/brave_rewards/resources/extension/brave_rewards/constants/rewards_panel_types.ts]: nit: maybe `ON_PUBLISHER_STATUS_REFRESHED`

Comment (jasonrsadler) [reply components/brave_rewards/resources/extension/brave_rewards/constants/rewards_panel_types.ts]: Done.

**Pull request merged**